File: 백준/Gold/2110. 공유기 설치/공유기 설치.py
import sys

class Solution:
    def __init__(self):
        self.input = sys.stdin.readline
        self.output = sys.stdout.write

    def determination(self, target, nums, c):
        cnt, last = 1, nums[0]
        for num in nums:
            if num - last < target:
                continue

            last = num
            cnt += 1
        return cnt >= c

    def run(self):
        nums = []
        n, c = map(int,self.input().split())
            
        for _ in range(n):
            nums.append(int(self.input()))

        nums.sort()
        left, right, ans = 0, 1000000000, 0

        while left <= right:
            mid = (left + right) // 2
            
            if self.determination(mid, nums, c):
                ans = mid
                left = mid + 1
                continue
            
            right = mid - 1

        self.output(f'{ans}\n')

s = Solution()
s.run()

File: 백준/Gold/2110. 공유기 설치/공유기 설치.java
import java.io.*;
import java.util.*;

public class Main {
    public static void main(String[] args) throws IOException{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        String[] temp = br.readLine().split(" ");
        int n = Integer.parseInt(temp[0]);
        int m = Integer.parseInt(temp[1]);

        List<Integer> nums = new ArrayList<>();

        for(int i = 0; i < n ; i++) nums.add(Integer.parseInt(br.readLine()));

        Collections.sort(nums);

        int left = 0;
        int right = 1_000_000_000;
        int mid = 0;
        int before = 0;
        int answer = 0;
        int temp_res = 1;

        while(left <= right){
            before = nums.get(0);
            temp_res = 1;
            mid = (left + right) / 2; 

            for(int j = 1 ; j < n ; j++){
                if(nums.get(j) - before < mid) continue;

                temp_res++;
                before = nums.get(j);
            }
            
            if(temp_res >= m){
                answer = mid;
                left = mid + 1;
                continue;
            }
            right = mid - 1;
        }
        System.out.println(answer);
        br.close();
    }
}